bsi_zoo/cross_val.py

```diff
@@ -0,0 +1,194 @@
+from scipy import linalg
+import numpy as np
+from sklearn.model_selection import GridSearchCV
+from sklearn.base import BaseEstimator, ClassifierMixin, clone
+from sklearn.model_selection import check_cv
+
+
+from .estimators import SpatialSolver
+
+
+def temporal_cv_metric(y, Sigma_Y):
+    """Compute the log-det Bregman divergence between two matrices.
+
+    It is based on the calculation of Gaussian negative log-likelihood.
+    Both matrices needs to be squared and have the same size.
+    It is given by:
+
+    B_inv = inv(B);
+    logdet_distance = trace(A*B_inv) - logdet(A*B_inv) - size(A,1)
+    """
+    # XXX Need improve w.r.t speed (Perhaps use the following conditions)
+    _, logdet = np.linalg.slogdet(Sigma_Y)
+    out = (np.linalg.norm(linalg.sqrtm(np.linalg.inv(Sigma_Y) @ y),
+                          ord='fro')**2 + logdet)
+    return out
+
+
+def _logdet(A):
+    """Compute the log det of a positive semidefinite matrix."""
+    from scipy import linalg
+    vals = linalg.eigvalsh(A)
+    # avoid negative (numerical errors) or zero (semi-definite matrix) values
+    tol = vals.max() * vals.size * np.finfo(np.float64).eps
+    vals = np.where(vals > tol, vals, tol)
+    return np.sum(np.log(vals))
+
+
+def logdet_bregman_div_distance_nll(y, Sigma_Y):
+    """Compute the Gaussian log likelihood of y given covariance Sigma_Y."""
+    # compute empirical covariance of the test set
+    precision = np.inv(Sigma_Y)
+    _, n_features = y.shape
+    log_like = -.5 * (y * (np.dot(y, precision))).sum(axis=1)
+    log_like -= .5 * (n_features * np.log(2. * np.pi) - _logdet(precision))
+    out = np.mean(log_like)
+    return out
+
+
+# XXX should be use this?
+# def logdet_bregman_div_distance(A, B):
+#     """Compute the log-det Bregman divergence between two matrices.
+
+#     It is based on the calculation of Gaussian negative log-likelihood.
+#     Both matrices needs to be squared and have the same size.
+
+#     B_inv = inv(B);
+#     logdet_distance = trace(A*B_inv) - logdet(A*B_inv) - size(A,1)
+
+#     Parameters
+#     ----------
+#     A : ndarray, shape (n_features, n_features)
+#         The first matrix.
+#     B : ndarray, shape (n_features, n_features)
+#         The second matrix
+
+#     Returns
+#     -------
+#     div : float
+#         The value of the Bregman divergence.
+#     """
+#     B_inv = np.linalg.inv(B)
+#     logdet_distance = (A * (np.dot(A, B_inv))).sum(axis=1)
+#     logdet_distance -= _logdet(A @ B_inv) + A.shape[0]
+#     out = np.mean(logdet_distance)
+#     return out
+
+
+# def logdet_bregman_div_distance_nll(y, Sigma_Y):
+#     """Compute the log-det Bregman divergence between
+#     two matrices based on the calculation of Gaussian
+#     negative log-likelihood.
+#     """
+#     log_lik = _gaussian_loglik_scorer(y, Sigma_Y)
+#     nll = -2 * log_lik
+#     return nll
```

Comment:
do we need the commented code above?

Reply:
No I will remove that. (I explained the reason above)